File: src/org/antlr/runtime/tree/CommonTreeNodeStream.java
/*
[The "BSD licence"]
Copyright (c) 2005-2006 Terence Parr
All rights reserved.

Redistribution and use in source and binary forms, with or without
modification, are permitted provided that the following conditions
are met:
1. Redistributions of source code must retain the above copyright
notice, this list of conditions and the following disclaimer.
2. Redistributions in binary form must reproduce the above copyright
notice, this list of conditions and the following disclaimer in the
documentation and/or other materials provided with the distribution.
3. The name of the author may not be used to endorse or promote products
derived from this software without specific prior written permission.

THIS SOFTWARE IS PROVIDED BY THE AUTHOR ``AS IS'' AND ANY EXPRESS OR
IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY DIRECT, INDIRECT,
INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT
NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
*/
package org.antlr.runtime.tree;

import org.antlr.runtime.Token;

import java.util.List;
import java.util.Iterator;
import java.util.ArrayList;

/** A buffered stream of tree nodes.  Nodes can be from a tree of ANY kind.
 *
 *  This node stream sucks all nodes out of the tree specified in
 *  the constructor during construction and makes pointers into
 *  the tree using an array of Object pointers. The stream necessarily
 *  includes pointers to DOWN and UP and EOF nodes.
 *
 *  This stream knows how to mark/release for backtracking.
 *
 *  This stream is most suitable for tree interpreters that need to
 *  jump around a lot or for tree parsers requiring speed (at cost of memory).
 *  There is some duplicated functionality here with UnBufferedTreeNodeStream
 *  but just in bookkeeping, not tree walking etc...
 *
 *  @see UnBufferedTreeNodeStream
 */
public class CommonTreeNodeStream implements TreeNodeStream {
	public static final int DEFAULT_INITIAL_BUFFER_SIZE = 100;
	public static final int INITIAL_CALL_STACK_SIZE = 10;

	protected static abstract class DummyTree extends BaseTree {
		public Tree dupNode() {return null;}
		public int getTokenStartIndex() { return 0; }
		public void setTokenStartIndex(int index) {}
		public int getTokenStopIndex() { return 0; }
		public void setTokenStopIndex(int index) {}
	}

	public static class NavDownNode extends DummyTree {
		public int getType() {return Token.DOWN;}
		public String getText() {return "DOWN";}
		public String toString() {return "DOWN";}
	}

	public static class NavUpNode extends DummyTree {
		public int getType() {return Token.UP;}
		public String getText() {return "UP";}
		public String toString() {return "UP";}
	}

	public static class EOFNode extends DummyTree {
		public int getType() {return Token.EOF;}
		public String getText() {return "EOF";}
		public String toString() {return "EOF";}
	}

	protected class StreamIterator implements Iterator {
		int i = 0;
		public boolean hasNext() {
			return i<nodes.size();
		}

		public Object next() {
			int current = i;
			i++;
			if ( current < nodes.size() ) {
				return nodes.get(current);
			}
			return EOF_NODE;
		}

		public void remove() {
			throw new RuntimeException("cannot remove nodes from stream");
		}
	}

	// all these navigation nodes are shared and hence they
	// cannot contain any line/column info

	public static final DummyTree DOWN = new NavDownNode();
	public static final DummyTree UP = new NavUpNode();
	public static final DummyTree EOF_NODE = new EOFNode();

	/** The complete mapping from stream index to tree node.
	 *  This buffer includes pointers to DOWN, UP, and EOF nodes.
	 *  It is built upon ctor invocation.  The elements are type
	 *  Object as we don't what the trees look like.
	 */
	protected List nodes;

	/** Pull nodes from which tree? */
	protected Object root;

	/** What tree adaptor was used to build these trees */
	TreeAdaptor adaptor;

	/** Reuse same DOWN, UP navigation nodes unless this is true */
	protected boolean uniqueNavigationNodes = false;

	/** The index into the nodes list of the current node (next node
	 *  to consume).
	 */
	protected int p = 0;

	/** Track the last mark() call result value for use in rewind(). */
	protected int lastMarker;

	/** Stack of indexes used for push/pop calls */
	protected int[] calls;

	/** Stack pointer for stack of indexes; -1 indicates empty.  Points
	 *  at next location to push a value.
	 */
	protected int _sp = -1;

	public CommonTreeNodeStream(Object tree) {
		this(new CommonTreeAdaptor(), tree);
	}

	public CommonTreeNodeStream(TreeAdaptor adaptor, Object tree) {
		this(adaptor, tree, DEFAULT_INITIAL_BUFFER_SIZE);
	}

	public CommonTreeNodeStream(TreeAdaptor adaptor, Object tree, int initialBufferSize) {
		this.root = tree;
		this.adaptor = adaptor;
		nodes = new ArrayList(initialBufferSize);
		fillBuffer(root); // starting at root, fill nodes buffer
	}

	/** Walk tree with depth-first-search and fill nodes buffer.
	 *  Don't do DOWN, UP nodes if its a list (p is isNil).
	 */
	protected void fillBuffer(Object p) {
		boolean nil = adaptor.isNil(p);
		if ( !nil ) {
			nodes.add(p); // add this node
		}
		// add DOWN node if p has children
		int n = adaptor.getChildCount(p);
		if ( !nil && n>0 ) {
			addNavigationNode(Token.DOWN);
		}
		// and now add all its children
		for (int c=0; c<n; c++) {
			Object child = adaptor.getChild(p,c);
			fillBuffer(child);
		}
		// add UP node if p has children
		if ( !nil && n>0 ) {
			addNavigationNode(Token.UP);
		}
	}

	/** As we flatten the tree, we use UP, DOWN nodes to represent
	 *  the tree structure.  When debugging we need unique nodes
	 *  so instantiate new ones when uniqueNavigationNodes is true.
	 */
	protected void addNavigationNode(final int ttype) {
		Object navNode = null;
		if ( ttype== Token.DOWN ) {
			if ( hasUniqueNavigationNodes() ) navNode = new CommonTreeNodeStream.NavDownNode();
			else navNode = CommonTreeNodeStream.DOWN;
		}
		else {
			if ( hasUniqueNavigationNodes() ) navNode = new CommonTreeNodeStream.NavUpNode();
			else navNode = CommonTreeNodeStream.UP;
		}
		nodes.add(navNode);
	}

	public Object LT(int k) {
		if ( k==0 ) {
			return null;
		}
		if ( k<0 ) {
			return LB(-k);
		}
		//System.out.print("LT(p="+p+","+k+")=");
		if ( (p+k-1) >= nodes.size() ) {
			return CommonTreeNodeStream.EOF_NODE;
		}
		return nodes.get(p+k-1);
	}

	/** Look backwards k nodes */
	protected Object LB(int k) {
		if ( k==0 ) {
			return null;
		}
		if ( (p-k)<0 ) {
			return null;
		}
		return nodes.get(p-k);
	}

	public Object getTreeSource() {
		return root;
	}

	public TreeAdaptor getTreeAdaptor() {
		return adaptor;
	}

	public boolean hasUniqueNavigationNodes() {
		return uniqueNavigationNodes;
	}

	public void setUniqueNavigationNodes(boolean uniqueNavigationNodes) {
		this.uniqueNavigationNodes = uniqueNavigationNodes;
	}

	public void consume() {
		p++;
	}

	public int LA(int i) {
		return adaptor.getType(LT(i));
	}

	public int mark() {
		lastMarker = index();
		return lastMarker;
	}

	public void release(int marker) {
		// no resources to release
	}

	public int index() {
		return p;
	}

	public void rewind(int marker) {
		seek(marker);
	}

	public void rewind() {
		seek(lastMarker);
	}

	public void seek(int index) {
		p = index;
	}

	/** Make stream jump to a new location, saving old location.
	 *  Switch back with pop().  I manage dyanmic array manually
	 *  to avoid creating Integer objects all over the place.
	 */
	public void push(int index) {
		if ( calls==null ) {
			calls = new int[INITIAL_CALL_STACK_SIZE];
		}
		else if ( (_sp+1)>=calls.length ) {
			int[] newStack = new int[calls.length*2];
			System.arraycopy(calls, 0, newStack, 0, calls.length-1);
			calls = newStack;
		}
		calls[++_sp] = p; // save current index
		seek(index);
	}

	/** Seek back to previous index saved during last push() call.
	 *  Return top of stack (return index).
	 */
	public int pop() {
		int ret = calls[_sp--];
		seek(ret);
		return ret;
	}

	public int size() {
		return nodes.size();
	}

	public Iterator iterator() {
		return new StreamIterator();
	}

	/** Used for testing, just return the token type stream */
	public String toString() {
		StringBuffer buf = new StringBuffer();
		for (int i = 0; i < nodes.size(); i++) {
			Object t = (Object) nodes.get(i);
			buf.append(" ");
			buf.append(adaptor.getType(t));
		}
		return buf.toString();
	}

	public String toString(Object start, Object stop) {
		if ( start==null || stop==null ) {
			return null;
		}
		// walk nodes looking for start
		Object t = null;
		int i = 0;
		for (; i < nodes.size(); i++) {
			t = nodes.get(i);
			if ( t==start ) {
				break;
			}
		}
		// now walk until we see stop, filling string buffer with text
		 StringBuffer buf = new StringBuffer();
		t = nodes.get(i);
		while ( t!=stop ) {
			String text = adaptor.getText(t);
			if ( text==null ) {
				text = " "+String.valueOf(adaptor.getType(t));
			}
			buf.append(text);
			i++;
			t = nodes.get(i);
		}
		// include stop node too
		String text = adaptor.getText(stop);
		if ( text==null ) {
			text = " "+String.valueOf(adaptor.getType(stop));
		}
		buf.append(text);
		return buf.toString();
	}
}

File: runtime/CSharp/Antlr3.Runtime/Antlr.Runtime.Tree/CommonTreeNodeStream.cs
namespace Antlr.Runtime.Tree
{
    using Antlr.Runtime;
    using System;
    using System.Collections;
    using System.Text;

    public class CommonTreeNodeStream : ITreeNodeStream, IIntStream, IEnumerable
    {
        protected int _sp;
        private ITreeAdaptor adaptor;
        protected int[] calls;
        public const int DEFAULT_INITIAL_BUFFER_SIZE = 100;
        protected object down;
        protected object eof;
        public static readonly IDictionary INDEX_ALL = new Hashtable();
        public const int INITIAL_CALL_STACK_SIZE = 10;
        protected int lastMarker;
        protected IList nodes;
        protected int p;
        protected internal object root;
        protected ITokenStream tokens;
        protected IDictionary tokenTypesToReverseIndex;
        protected IDictionary tokenTypeToStreamIndexesMap;
        protected bool uniqueNavigationNodes;
        protected object up;

        public CommonTreeNodeStream(object tree) : this(new CommonTreeAdaptor(), tree)
        {
        }

        public CommonTreeNodeStream(ITreeAdaptor adaptor, object tree) : this(adaptor, tree, 100)
        {
        }

        public CommonTreeNodeStream(ITreeAdaptor adaptor, object tree, int initialBufferSize)
        {
            this.uniqueNavigationNodes = false;
            this.p = -1;
            this._sp = -1;
            this.tokenTypesToReverseIndex = null;
            this.root = tree;
            this.adaptor = adaptor;
            this.nodes = new ArrayList(initialBufferSize);
            this.down = adaptor.Create(2, "DOWN");
            this.up = adaptor.Create(3, "UP");
            this.eof = adaptor.Create(Token.EOF, "EOF");
        }

        protected void AddNavigationNode(int ttype)
        {
            object down = null;
            if (ttype == 2)
            {
                if (this.HasUniqueNavigationNodes)
                {
                    down = this.adaptor.Create(2, "DOWN");
                }
                else
                {
                    down = this.down;
                }
            }
            else if (this.HasUniqueNavigationNodes)
            {
                down = this.adaptor.Create(3, "UP");
            }
            else
            {
                down = this.up;
            }
            this.nodes.Add(down);
        }

        public virtual void Consume()
        {
            if (this.p == -1)
            {
                this.FillBuffer();
            }
            this.p++;
        }

        protected void FillBuffer()
        {
            this.FillBuffer(this.root);
            this.p = 0;
        }

        protected void FillBuffer(object t)
        {
            bool flag = this.adaptor.IsNil(t);
            if (!flag)
            {
                this.nodes.Add(t);
                this.FillReverseIndex(t, this.nodes.Count - 1);
            }
            int childCount = this.adaptor.GetChildCount(t);
            if (!(flag || (childCount <= 0)))
            {
                this.AddNavigationNode(2);
            }
            for (int i = 0; i < childCount; i++)
            {
                object child = this.adaptor.GetChild(t, i);
                this.FillBuffer(child);
            }
            if (!(flag || (childCount <= 0)))
            {
                this.AddNavigationNode(3);
            }
        }

        protected void FillReverseIndex(object node, int streamIndex)
        {
            if (this.tokenTypesToReverseIndex != null)
            {
                if (this.tokenTypeToStreamIndexesMap == null)
                {
                    this.tokenTypeToStreamIndexesMap = new Hashtable();
                }
                int nodeType = this.adaptor.GetNodeType(node);
                if ((this.tokenTypesToReverseIndex == INDEX_ALL) || this.tokenTypesToReverseIndex.Contains(nodeType))
                {
                    ArrayList list = (ArrayList) this.tokenTypeToStreamIndexesMap[nodeType];
                    if (list == null)
                    {
                        list = new ArrayList();
                        list.Add(streamIndex);
                        this.tokenTypeToStreamIndexesMap[nodeType] = list;
                    }
                    else if (!list.Contains(streamIndex))
                    {
                        list.Add(streamIndex);
                    }
                }
            }
        }

        public object Get(int i)
        {
            if (this.p == -1)
            {
                this.FillBuffer();
            }
            return this.nodes[i];
        }

        public IEnumerator GetEnumerator()
        {
            if (this.p == -1)
            {
                this.FillBuffer();
            }
            return new CommonTreeNodeStreamEnumerator(this);
        }

        public int GetNodeIndex(object node)
        {
            if (this.tokenTypeToStreamIndexesMap == null)
            {
                return this.GetNodeIndexLinearly(node);
            }
            int nodeType = this.adaptor.GetNodeType(node);
            ArrayList list = (ArrayList) this.tokenTypeToStreamIndexesMap[nodeType];
            if (list == null)
            {
                return this.GetNodeIndexLinearly(node);
            }
            for (int i = 0; i < list.Count; i++)
            {
                int num3 = (int) list[i];
                if (this.Get(num3) == node)
                {
                    return num3;
                }
            }
            return -1;
        }

        protected int GetNodeIndexLinearly(object node)
        {
            if (this.p == -1)
            {
                this.FillBuffer();
            }
            for (int i = 0; i < this.nodes.Count; i++)
            {
                object obj2 = this.nodes[i];
                if (obj2 == node)
                {
                    return i;
                }
            }
            return -1;
        }

        public virtual int Index()
        {
            return this.p;
        }

        public virtual int LA(int i)
        {
            return this.adaptor.GetNodeType(this.LT(i));
        }

        protected object LB(int k)
        {
            if (k == 0)
            {
                return null;
            }
            if ((this.p - k) < 0)
            {
                return null;
            }
            return this.nodes[this.p - k];
        }

        public object LT(int k)
        {
            if (this.p == -1)
            {
                this.FillBuffer();
            }
            if (k == 0)
            {
                return null;
            }
            if (k < 0)
            {
                return this.LB(-k);
            }
            if (((this.p + k) - 1) >= this.nodes.Count)
            {
                return this.eof;
            }
            return this.nodes[(this.p + k) - 1];
        }

        public virtual int Mark()
        {
            if (this.p == -1)
            {
                this.FillBuffer();
            }
            this.lastMarker = this.Index();
            return this.lastMarker;
        }

        public int Pop()
        {
            int index = this.calls[this._sp--];
            this.Seek(index);
            return index;
        }

        public void Push(int index)
        {
            if (this.calls == null)
            {
                this.calls = new int[10];
            }
            else if ((this._sp + 1) >= this.calls.Length)
            {
                int[] destinationArray = new int[this.calls.Length * 2];
                Array.Copy(this.calls, 0, destinationArray, 0, this.calls.Length);
                this.calls = destinationArray;
            }
            this.calls[++this._sp] = this.p;
            this.Seek(index);
        }

        public virtual void Release(int marker)
        {
        }

        public void ReverseIndex(IDictionary tokenTypes)
        {
            this.tokenTypesToReverseIndex = tokenTypes;
        }

        public void ReverseIndex(int tokenType)
        {
            if (this.tokenTypesToReverseIndex == null)
            {
                this.tokenTypesToReverseIndex = new Hashtable();
            }
            else if (this.tokenTypesToReverseIndex == INDEX_ALL)
            {
                return;
            }
            this.tokenTypesToReverseIndex.Add(tokenType, tokenType);
        }

        public void Rewind()
        {
            this.Seek(this.lastMarker);
        }

        public virtual void Rewind(int marker)
        {
            this.Seek(marker);
        }

        public virtual void Seek(int index)
        {
            if (this.p == -1)
            {
                this.FillBuffer();
            }
            this.p = index;
        }

        public virtual int Size()
        {
            if (this.p == -1)
            {
                this.FillBuffer();
            }
            return this.nodes.Count;
        }

        public override string ToString()
        {
            if (this.p == -1)
            {
                this.FillBuffer();
            }
            StringBuilder builder = new StringBuilder();
            for (int i = 0; i < this.nodes.Count; i++)
            {
                object t = this.nodes[i];
                builder.Append(" ");
                builder.Append(this.adaptor.GetNodeType(t));
            }
            return builder.ToString();
        }

        public virtual string ToString(object start, object stop)
        {
            string nodeText;
            if ((start == null) || (stop == null))
            {
                return null;
            }
            if (this.p == -1)
            {
                this.FillBuffer();
            }
            Console.Out.WriteLine("stop: " + stop);
            if (start is CommonTree)
            {
                Console.Out.Write("ToString: " + ((CommonTree) start).Token + ", ");
            }
            else
            {
                Console.Out.WriteLine(start);
            }
            if (stop is CommonTree)
            {
                Console.Out.WriteLine(((CommonTree) stop).Token);
            }
            else
            {
                Console.Out.WriteLine(stop);
            }
            if (this.tokens != null)
            {
                int tokenStartIndex = this.adaptor.GetTokenStartIndex(start);
                int tokenStopIndex = this.adaptor.GetTokenStopIndex(stop);
                if (this.adaptor.GetNodeType(stop) == 3)
                {
                    tokenStopIndex = this.adaptor.GetTokenStopIndex(start);
                }
                else if (this.adaptor.GetNodeType(stop) == Token.EOF)
                {
                    tokenStopIndex = this.Size() - 2;
                }
                return this.tokens.ToString(tokenStartIndex, tokenStopIndex);
            }
            object t = null;
            int num3 = 0;
            while (num3 < this.nodes.Count)
            {
                t = this.nodes[num3];
                if (t == start)
                {
                    break;
                }
                num3++;
            }
            StringBuilder builder = new StringBuilder();
            for (t = this.nodes[num3]; t != stop; t = this.nodes[num3])
            {
                nodeText = this.adaptor.GetNodeText(t);
                if (nodeText == null)
                {
                    nodeText = " " + this.adaptor.GetNodeType(t);
                }
                builder.Append(nodeText);
                num3++;
            }
            nodeText = this.adaptor.GetNodeText(stop);
            if (nodeText == null)
            {
                nodeText = " " + this.adaptor.GetNodeType(stop);
            }
            builder.Append(nodeText);
            return builder.ToString();
        }

        public bool HasUniqueNavigationNodes
        {
            get
            {
                return this.uniqueNavigationNodes;
            }
            set
            {
                this.uniqueNavigationNodes = value;
            }
        }

        public virtual ITokenStream TokenStream
        {
            get
            {
                return this.tokens;
            }
            set
            {
                this.tokens = value;
            }
        }

        public ITreeAdaptor TreeAdaptor
        {
            get
            {
                return this.adaptor;
            }
        }

        public virtual object TreeSource
        {
            get
            {
                return this.root;
            }
        }

        protected sealed class CommonTreeNodeStreamEnumerator : IEnumerator
        {
            private object _currentItem;
            private int _index;
            private CommonTreeNodeStream _nodeStream;

            internal CommonTreeNodeStreamEnumerator()
            {
            }

            internal CommonTreeNodeStreamEnumerator(CommonTreeNodeStream nodeStream)
            {
                this._nodeStream = nodeStream;
                this.Reset();
            }

            public bool MoveNext()
            {
                if (this._index >= this._nodeStream.nodes.Count)
                {
                    int num = this._index;
                    this._index++;
                    if (num < this._nodeStream.nodes.Count)
                    {
                        this._currentItem = this._nodeStream.nodes[num];
                    }
                    this._currentItem = this._nodeStream.eof;
                    return true;
                }
                this._currentItem = null;
                return false;
            }

            public void Reset()
            {
                this._index = 0;
                this._currentItem = null;
            }

            public object Current
            {
                get
                {
                    if (this._currentItem == null)
                    {
                        throw new InvalidOperationException("Enumeration has either not started or has already finished.");
                    }
                    return this._currentItem;
                }
            }
        }
    }
}

